### PR TITLE
Better local metadata comparison with ZooKeeper metadata

### PR DESCRIPTION
**Pull request by sunny19930321**
Changelog category (leave one):
- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry:
Fix parsing ZK metadata: now metadata from zookeeper compared with local metadata in canonical form.

### CODE REVIEW

**Comment by alesapin** [comment]
@mergifyio update

File: src/Common/IFactoryWithAliases.h [hunk 1]
```diff
@@ -120,8 +120,12 @@ class IFactoryWithAliases : public IHints<2, IFactoryWithAliases<ValueType>>
     const String & getCanonicalNameIfAny(const String & name) const
     {
         auto it = case_insensitive_name_mapping.find(Poco::toLower(name));
-        if (it != case_insensitive_name_mapping.end())
-            return it->second;
+        if (it != case_insensitive_name_mapping.end()) {
+            if (it->first != name)
```
[the remark this answers is not in the document]
Comment (sunny19930321) [reply src/Common/IFactoryWithAliases.h:124]: thanks for your advice, there are two solutions, the first is to fix at the transformation layer, and the second is to fix at the point where the exception is thrown.I'll change it as you suggested.

[the remark this answers is not in the document]
Comment (sunny19930321) [reply src/Common/IFactoryWithAliases.h]: @alesapin cc

**Comment by alesapin** [comment]
@Mergifyio update

**Review by alesapin — APPROVED** [review]
I think it's quite hard to add tests, but change is clear. LGTM.

**Comment by tavplubix** [comment]
@sunny19930321, @alesapin, looks like this PR breaks compatibility, please check https://github.com/ClickHouse/ClickHouse/issues/46834

**Comment by sunny19930321** [comment]
> @sunny19930321, @alesapin, looks like this PR breaks compatibility, please check #46834

the problem doesn't seem to be due to pr. when zookeeper gets the TTL, it doesn't get it

**Pull request merged**